zio-http/src/main/scala/zhttp/http/Headers.scala

```diff
@@ -15,59 +16,95 @@ import scala.jdk.CollectionConverters._
+ * NOTE: Generic operators that are not specific to `Headers` should not be
+ * defined here. A better place would be one of the traits extended by
+ * `HeaderExtension`.
+ */
+sealed trait Headers extends HeaderExtension[Headers] {
   self =>
 
-  def ++(other: Headers): Headers = self.combine(other)
+  def ++(other: Headers): Headers = HeadersCons(self, other)
 
-  def combine(other: Headers): Headers = Headers(self.toChunk ++ other.toChunk)
+  def combine(other: Headers): Headers = HeadersCons(self, other)
 
-  def combineIf(cond: Boolean)(other: Headers): Headers = if (cond) Headers(self.toChunk ++ other.toChunk) else self
+  def combineIf(cond: Boolean)(other: Headers): Headers = if (cond)
+    HeadersCons(self, other)
+  else self
 
   override def headers: Headers = self
 
-  def modify(f: Header => Header): Headers = Headers(toChunk.map(f(_)))
-
-  def toList: List[(String, String)] = toChunk.map { case (name, value) => (name.toString, value.toString) }.toList
+  def modify(f: Header => Header): Headers = ModifyHeaders(f, self)
+
+  def toList: List[(String, String)] = self match {
+    case EmptyHeaders                    => List()
+    case HeadersCons(a, b)               => a.toList ++ b.toList
+    case HeadersFromChunk(chunk)         => chunk.map(h => (h._1.toString, h._2.toString)).toList
+    case HeadersFromHttp(headers)        =>
+      headers
+        .iteratorCharSequence()
+        .asScala
+        .map(h => (h.getKey.toString, h.getValue.toString))
+        .toList
+    case ModifyHeaders(modify, headers)  => headers.toList.map(modify).map(h => (h._1.toString, h._2.toString))
+    case UpdatedHeaders(update, headers) =>
+      update(headers).toList
+  }
 
-  override def updateHeaders(update: Headers => Headers): Headers = update(self)
+  override def updateHeaders(update: Headers => Headers): Headers =
+    UpdatedHeaders(update, this)
 
   def when(cond: Boolean): Headers = if (cond) self else Headers.empty
 
   /**
    * Converts a Headers to [io.netty.handler.codec.http.HttpHeaders]
    */
   private[zhttp] def encode: HttpHeaders =
-    self.toList
-      .foldLeft[HttpHeaders](new CombinedHttpHeaders(true)) { case (headers, entry) =>
-        headers.add(entry._1, entry._2)
-      }
+    self.toList.foldLeft[HttpHeaders](new CombinedHttpHeaders(true)) { case (headers, entry) =>
+      headers.add(entry._1, entry._2)
+    }
+}
 
+case object EmptyHeaders                                              extends Headers {
+  override def header(headerName: CharSequence): Option[Header] = None
+}
+case class HeadersFromChunk(value: Chunk[Header])                     extends Headers {
+  override def header(headerName: CharSequence): Option[Header] =
+    value.find(h => contentEqualsIgnoreCase(h._1, headerName))
+}
+case class HeadersFromHttp(value: HttpHeaders)                        extends Headers {
+  override def header(headerName: CharSequence): Option[Header] =
+    Option(value.get(headerName)).map((headerName, _))
```

Comment:
Can improve perf further by dropping the `map` and using the basic `if` check.

Reply:
Ok, updated it in the laatest commit (avoiding value.contains because it would internally call 'get' twice)